### PR TITLE
Integrate new data util into Tidepool web [WEB-15, WEB-16, WEB-18, WEB-20, WEB-22, WEB-24]

### PR DESCRIPTION
**Pull request by cbwebdevelopment**
Part of new Data Preprocessing feature [WEB-4]

Addresses [WEB-15], [WEB-16], [WEB-18], [WEB-20], [WEB-22], [WEB-24]

Related PRs:
tidepool-org/viz#177
tidepool-org/tideline#388

**NOTE**: I've requested a review of this PR prior to completing all of the unit tests as I believe it will take a while to review, and I'm looking to expediate things a little.

The unit tests that remain are for the `patientData.js` file, and the redux-related tests.  I will hopefully have these all done shortly and will ping for review of those tests when ready.

[WEB-15]: https://tidepool.atlassian.net/browse/WEB-15
[WEB-18]: https://tidepool.atlassian.net/browse/WEB-18
[WEB-20]: https://tidepool.atlassian.net/browse/WEB-20
[WEB-22]: https://tidepool.atlassian.net/browse/WEB-22
[WEB-24]: https://tidepool.atlassian.net/browse/WEB-24
[WEB-16]: https://tidepool.atlassian.net/browse/WEB-16

[WEB-4]: https://tidepool.atlassian.net/browse/WEB-4

### CODE REVIEW

**Comment by cbwebdevelopment** [comment]
@krystophv Just an update on the remaining unit tests:

Everything is now done except for 8 new methods I still need to write tests for in patientdata.js

File: test/unit/pages/patientdata.test.js [hunk 1]
```diff
@@ -1121,2206 +1246,2280 @@ describe('PatientData', function () {
...
+              wrapper.setProps({
+                ...notAddingDataProps,
+                ...fetchingDataProps,
+              });
 
-        let newDateRange = [
-          moment(instance.props.fetchedPatientDataRange.start).subtract(1, 'milliseconds').toISOString(),
-          dateRange[1],
-        ];
+
```
Comment (krystophv) [inline test/unit/pages/patientdata.test.js]: Intentionally skipping?

File: test/unit/pages/patientdata.test.js [hunk 2]
```diff
@@ -3397,34 +3596,36 @@ describe('PatientData', function () {
...
       const wrapper = shallow(<PatientData.WrappedComponent {...defaultProps} />);
       const instance = wrapper.instance();
+      wrapper.setState({timePrefs: { timezoneAware: true, timezoneName: 'utc' } })
 
-      const chartPrefs = { basics: 'basics prefs' };
-
-      wrapper.setState({ chartPrefs })
-      instance.dataUtil = new DataUtilStub();
+      instance.updateChart = sinon.stub();
+      instance.getMostRecentDatumTimeByChartType = sinon.stub().returns('2019-11-27T12:00:00.000Z');
```
Comment (krystophv) [inline test/unit/pages/patientdata.test.js]: Should this be explicitly checked for a call later?

File: test/unit/pages/patientdata.test.js [hunk 3]
```diff
@@ -3448,31 +3648,33 @@ describe('PatientData', function () {
...
+      instance.getChartEndpoints = sinon.stub().returns('endpoints stub');
 
       instance.handleSwitchToDaily();
-      expect(instance.dataUtil._chartPrefs).to.equal('daily prefs');
+      console.log(instance.getChartEndpoints.getCall(0).args)
```
Comment (krystophv) [inline test/unit/pages/patientdata.test.js]: Console log left in? (there's a couple more lower down as well)

File: test/unit/pages/patientdata.test.js [hunk 4]
```diff
@@ -3814,6 +4012,22 @@ describe('PatientData', function () {
...
+
+      it('should map working.updatingDatum to updatingDatum', () => {
+        expect(result.updatingDatum).to.equal(state.working.updatingDatum);
+      });
+
+      it('should map working.queryingData to queryingData', () => {
+        expect(result.queryingData).to.equal(state.working.queryingData);
+      });
+
+      it('should map blip.data to data', () => {
```
Comment (krystophv) [inline test/unit/pages/patientdata.test.js]: `blip.data` -> `working.data`? (and another instance lower)

File: app/components/chart/basics.js [hunk 5]
```diff
@@ -145,14 +121,15 @@ class Basics extends Component {
...
+          bgUnits={_.get(this.props, 'data.bgPrefs', {}).bgUnits}
+          data={this.props.data}
           onSelectDay={this.handleSelectDay}
           patient={this.props.patient}
-          patientData={this.props.patientData}
           permsOfLoggedInUser={this.props.permsOfLoggedInUser}
-          timePrefs={this.props.timePrefs}
-          updateBasicsData={this.props.updateBasicsData}
+          timePrefs={_.get(this.props, 'data.timePrefs', {})}
+          // updateBasicsData={this.props.updateBasicsData}
```
[the remark this answers is not in the document]
Comment (cbwebdevelopment) [reply app/components/chart/basics.js]: Yep.  Will do.

File: app/pages/patientdata/patientdata.js [hunk 6]
```diff
@@ -1471,34 +1706,37 @@ export function mapStateToProps(state, props) {
...
     pdf: state.blip.pdf,
-    viz: state.viz,
+    data: state.blip.data,
   };
 }
 
 let mapDispatchToProps = dispatch => bindActionCreators({
-  addPatientNote: actions.sync.addPatientNote,
-  clearPatientData: actions.sync.clearPatientData,
+  dataWorkerAddDataRequest: actions.worker.dataWorkerRemoveDataRequest,
```
[the remark this answers is not in the document]
Comment (cbwebdevelopment) [reply app/pages/patientdata/patientdata.js]: Yes! Had me wondering how this was working, but it seems that `dataWorkerAddDataRequest` is not actually being used, so I can delete this line.

File: test/unit/pages/patientdata.test.js [hunk 7]
```diff
@@ -1121,2206 +1246,2280 @@ describe('PatientData', function () {
...
-      wrapper.instance().getWrappedInstance().setState({ chartType: 'daily', processingData: false, processedPatientData: true });
-      wrapper.update();
+    it('should return the latest datum time for bgLog', () => {
+      // should return the smbg normalTime
+      expect(instance.getMostRecentDatumTimeByChartType(undefined, 'bgLog')).to.equal(3);
+    });
 
-      expect(elem.generatePDF.callCount).to.equal(0);
+    it('should return the latest datum time for trends', () => {
+      // should return the smbg normalTime
```
[the remark this answers is not in the document]
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: 👍

File: test/unit/pages/patientdata.test.js [hunk 8]
```diff
@@ -1121,2206 +1246,2280 @@ describe('PatientData', function () {
...
-      sinon.assert.callCount(bgLogFilterStub, 1);
-      sinon.assert.callCount(props.generatePDFRequest, 1);
+        instance.updateChart(defaultChartType, '2019-11-28T00:00:00.000Z', defaultEndpoints, defaultOpts);
+      });
 
-      sinon.assert.callCount(pickSpy, 2);
-      assert(dailyFilterStub.calledBefore(bgLogFilterStub));
-      expect(pickSpy.firstCall.lastArg).to.have.members(['basal', 'bolus', 'cbg', 'food', 'message', 'smbg', 'upload']);
-      expect(pickSpy.secondCall.lastArg).to.have.members(['smbg']);
+      it('should set `chartType`, `transitioningChartType`, and `mostRecentDatetimelocation` state', () => {
```
[the remark this answers is not in the document]
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: Yep.  Fixing.

File: test/unit/pages/patientdata.test.js [hunk 9]
```diff
@@ -1121,2206 +1246,2280 @@ describe('PatientData', function () {
...
-            pumpSettings: {},
-          },
-        },
-        printOpts: {
-          numDays: {
-            daily: 6,
-          },
-        },
-      };
+      it('should set `chartType`, `transitioningChartType`, and `mostRecentDatetimelocation` state', () => {
```
[the remark this answers is not in the document]
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: 👍

File: test/unit/pages/patientdata.test.js [hunk 10]
```diff
@@ -1121,2206 +1246,2280 @@ describe('PatientData', function () {
...
-        data.basics = { stats: 'stubbed basics stats' };
-        return data;
+        instance.updateChart(defaultChartType, defaultDatetimeLocation, defaultEndpoints, defaultOpts);
       });
 
-      instance.generatePDF(props, state);
-
-      sinon.assert.callCount(props.generatePDFRequest, 1);
-      sinon.assert.callCount(instance.generatePDFStats, 1);
+      it('should set `chartType`, `transitioningChartType`, and `mostRecentDatetimelocation` state', () => {
```
[the remark this answers is not in the document]
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: :) Good catch

File: test/unit/pages/patientdata.test.js (hunk 1 above)
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: Yes.  Will be re-enabled with upcoming PDF integration PR

File: test/unit/pages/patientdata.test.js (hunk 3 above)
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: Removing...

File: test/unit/pages/patientdata.test.js (hunk 4 above)
Comment (cbwebdevelopment) [reply test/unit/pages/patientdata.test.js]: Yep.  Fixed both occurrences

**Pull request merged**